From 25c79cce39f40ba75def8602b05e4208d1d83b95 Mon Sep 17 00:00:00 2001
From: Akhil Goyal <akhil.goyal@nxp.com>
Date: Thu, 11 Aug 2016 16:45:39 +0530
Subject: [PATCH 053/166] examples/ipsec-secgw: update mbuf packet type

Packet_type should be updated to remove/add L4 type for
encrypted/decrypted packet

Signed-off-by: Akhil Goyal <akhil.goyal@nxp.com>
---
 examples/ipsec-secgw/ipip.h | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/examples/ipsec-secgw/ipip.h b/examples/ipsec-secgw/ipip.h
index 93393d527f2b..469b044ed054 100644
--- a/examples/ipsec-secgw/ipip.h
+++ b/examples/ipsec-secgw/ipip.h
@@ -56,6 +56,7 @@ ipip_outbound(struct rte_mbuf *m, uint32_t offset, uint32_t is_ipv6,
 	if (inip4->ip_v == IPVERSION) {
 		/* XXX This should be done by the forwarding engine instead */
 		inip4->ip_ttl -= 1;
+		inip4->ip_sum += 1;
 		ds_ecn = inip4->ip_tos;
 	} else {
 		inip6 = (struct ip6_hdr *)inip4;
@@ -86,6 +87,7 @@ ipip_outbound(struct rte_mbuf *m, uint32_t offset, uint32_t is_ipv6,
 
 	offset += sizeof(struct ip);
 	outip4 = (struct ip *)rte_pktmbuf_prepend(m, offset);
+	memset(outip4, 0, sizeof(struct ip));
 
 	RTE_ASSERT(outip4 != NULL);
 
@@ -103,7 +105,7 @@ ipip_outbound(struct rte_mbuf *m, uint32_t offset, uint32_t is_ipv6,
 
 	outip4->ip_src.s_addr = src->ip.ip4;
 	outip4->ip_dst.s_addr = dst->ip.ip4;
-
+	m->packet_type &= ~RTE_PTYPE_L4_MASK;
 	return outip4;
 }
 
@@ -169,6 +171,12 @@ ipip_inbound(struct rte_mbuf *m, uint32_t offset)
 			ip4_ecn_setup(inip4);
 		/* XXX This should be done by the forwarding engine instead */
 		inip4->ip_ttl -= 1;
+		inip4->ip_sum += 1;
+		m->packet_type &= ~RTE_PTYPE_L4_MASK;
+		if (inip4->ip_p == IPPROTO_UDP)
+			m->packet_type |= RTE_PTYPE_L4_UDP;
+		else if (inip4->ip_p == IPPROTO_TCP)
+			m->packet_type |= RTE_PTYPE_L4_TCP;
 	} else {
 		inip6 = (struct ip6_hdr *)inip4;
 		if (set_ecn)
-- 
2.14.3

